### ISSUE TITLE
QT Android app frozen on MutexImpl::lockImpl() (phthread_mutex_lock)

### ISSUE THREAD

**Issue by unimaxis** [pocoproject/poco#2498]
#### Expected behavior
There is a project on QT for Android. The Poco library was compiled using clang as dynamic, used the toolchain from Android NDKr17. The problem is that the incoming connection to the nttp server hangs. Under debugging, it turned out that the execution of the code was stalled in method
"MutexImpl::lockImpl()" on phthread_mutex_lock
The screen shows a problem under the debugger.
http://joxi.ru/YmEDB8Ks04LK3r

#### Actual behavior
build POCO:
/home/maxim/dev/Android/cmake/3.6.4111459/bin/cmake -H/home/maxim/work/kebox/poco-1.9.1 -B/home/maxim/work/kebox/poco-1.9.1/cmake-build -G'Android Gradle - Ninja' -DANDROID_ABI=armeabi-v7a -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE=/home/maxim/dev/Android/ndk-bundle/build/cmake/android.toolchain.cmake -DANDROID_PLATFORM=android-21 -DCMAKE_MAKE_PROGRAM=/home/maxim/dev/Android/cmake/3.6.4111459/bin/ninja -DANDROID_STL="c++_static" -DANDROID_CPP_FEATURES="exceptions rtti" -DENABLE_TESTS=OFF -DBUILD_SHARED_LIBS=OFF && /home/maxim/dev/Android/cmake/3.6.4111459/bin/cmake --build /home/maxim/work/kebox/poco-1.9.1/cmake-build --target all

build QT for Android with clang support:
./configure -confirm-license -opensource -xplatform android-clang -nomake tests -nomake examples -android-ndk $ndk_path -android-sdk $sdk_path -android-ndk-host linux-x86_64 -android-toolchain-version 4.9 -no-warnings-are-errors -no-pkg-config -qt-zlib -qt-freetype -android-arch arm64-v8a -android-ndk-platform android-21 -android-arch armeabi-v7a -prefix $prefix -skip qttranslations -skip qtserialport && make -j3 && make install

#### Steps to reproduce the problem
Simple example of httprest server
[PocoAndroidTest.zip](https://github.com/pocoproject/poco/files/2464693/PocoAndroidTest.zip)

curl -v http://device_ip:8080/system/version

#### POCO version
version 1.9.1

#### Compiler and version
Android v6.0
Android NDK r17
GCC 4.9 from NDK
Clang v 6.0.2 from NDK toolchacin

#### Operating system and version
Ubuntu 16.04 x64

**Comment by unimaxis** [comment]
I tried use of POCO v.1.8.0, the situation is similar but slightly better. Http requests in the test application are processed through.

**Comment by Bjoe** [comment]
Hi again @unimaxis :-)

I try to help you and get your PocoAndroidTest example, but I'm confused.
You said:

> The Poco library was compiled using clang as dynamic

But I see in your build POCO:
```
-DBUILD_SHARED_LIBS=OFF
```
also I found in the PocoAndroidTest.pro Line 43
```
/opt/keboxcontrib/lib-armeabi-v7/libPocoNetd.a ...
```
So Poco is compiled as static.

I see you build Qt from the source, because of clang ?! Which STL the Qt build is used? Can you send the output from objdump -p libQt5Core.so.

Before you build Qt from the source, have you try to use the build from Qt? As default, Qt is using the libgnustl as shared, so you should change the following build parameters on the poco build:
```
-DANDROID_STL="gnustl_shared"
```
For more information about the ANDROID parameters see: https://developer.android.com/ndk/guides/cmake and https://developer.android.com/ndk/guides/cpp-support

What I also noticed in PocoAndroidTest.pro Line 17:
```
DEFINES += POCO_ANDROID
```
I didn't understand why this is set? You mean maybe this: POCO_OS_ANDROID but this is set via https://github.com/pocoproject/poco/blob/85b9fe7a21b132bab8b9ca17003f004fded9430a/Foundation/include/Poco/Platform.h#L67

But I think there are some missing DEFINES, but this depends how Poco is build with the NDK. Maybe you send the cmake output from your Poco build.

Yes, and maybe I miss understand you, but I didn't get the full picture of your build environment and problem yet. I need more input :-)

**Comment by unimaxis** [comment]
> also I found in the PocoAndroidTest.pro Line 43

Hi @Bjoe ) 
I used different build options. I will describe in points the current environment:
1. Poco library build with flags -DANDROID_STL="gnustl_shared" and -DANDROID_STL="gnustl_shared"
2. Used precompiller flag #define POCO_ANDROID. 
A recommendation for its use is provided in the documentation https://pocoproject.org/docs/99300-AndroidPlatformNotes.html

I will try to build the project using the POCO_OS_ANDROID flag

Objdump of  libQt5Core.so
[libQt5Core.obj.dump.txt](https://github.com/pocoproject/poco/files/2477603/libQt5Core.obj.dump.txt)

**Comment by unimaxis** [comment]
@Bjoe 
Building a project with #define POCO_OS_ANDROID did not solve the problem. The request to the http server periodically freezes

**Comment by Bjoe** [comment]
No I didn't mean you should add `POCO_OS_ANDROID` as define, but I think you should add `POCO_STATIC` when you build Poco as static lib or `POCO_ENABLE_CPP11` when your compiler has support for C++11 .... I can give you the defines when you give me the cmake output of the poco build.

> A recommendation for its use is provided in the documentation https://pocoproject.org/docs/99300-AndroidPlatformNotes.html

Ah, ok. I should update this document. I didn't know there is also a document how to build on android....

> Objdump of libQt5Core.so

I see 
`NEEDED               libc++_shared.so` you build Qt with shared libc++ .... you should also build poco with that library. Please change the cmake parameter from: `-DANDROID_STL="c++_static"` to `-DANDROID_STL="c++_shared`

Btw. you will see also `NEEDED               libstdc++.so` .... this is a bug in the clang from Android NDK! I should create a bug report in Google for that :-) ....

**Comment by unimaxis** [comment]
@Bjoe 
I am building the POCO from dev branch and problem is solved. Need to test well.

**Comment by Bjoe** [comment]
Ok, I will investigate the problem. Let me figured out, but it will take some time!

**Comment by Bjoe** [comment]
Hmm, maybe it is the same issue like #2513 ... but this is only a assumption from me. I will verify this.

**Comment by niparx** [comment]
Same. I'm using [1.9.0 release](https://github.com/pocoproject/poco/releases/tag/poco-1.9.0-release). It freezes on `Poco::Data::SQLite::Connector::registerConnector()` call, which is, at the end, here: [Mutex_POSIX.h](https://github.com/pocoproject/poco/blob/13eda19cd1c2685113b535da7bf36aa88dfdaa5d/Foundation/include/Poco/Mutex_POSIX.h#L59). Sometimes, it passes this call. But ,in the end, freezes. Cause is on same line (`Mutex_POSIX.h:59`). Happens, when i'm trying to  create execution statement.

We were used 1.7.8 before, ant that version did not have such issue. Looks like we'll continue use same 1.7.8.

P.S. Build environment:
Android NDK: r15e
CMake: 3.14.4
Poco was build as static library. Toolchain setup was done as it is described [here](https://cmake.org/cmake/help/v3.14/manual/cmake-toolchains.7.html#cross-compiling-for-android)

**Comment by niparx** [comment]
Can confirm, that I have no problems on branch `poco-1.9.1`. Since it is not released, I'll wait for next `1.10` release, which will happens in next few weeks (as I understood).
